presto-thrift-connector/src/test/java/com/facebook/presto/connector/thrift/clientproviders/MockConnectedThriftServiceProvider.java

```diff
@@ -0,0 +1,155 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.connector.thrift.clientproviders;
+
+import com.facebook.presto.connector.thrift.api.PrestoThriftId;
+import com.facebook.presto.connector.thrift.api.PrestoThriftNullableColumnSet;
+import com.facebook.presto.connector.thrift.api.PrestoThriftNullableSchemaName;
+import com.facebook.presto.connector.thrift.api.PrestoThriftNullableTableMetadata;
+import com.facebook.presto.connector.thrift.api.PrestoThriftNullableToken;
+import com.facebook.presto.connector.thrift.api.PrestoThriftPageResult;
+import com.facebook.presto.connector.thrift.api.PrestoThriftSchemaTableName;
+import com.facebook.presto.connector.thrift.api.PrestoThriftServiceException;
+import com.facebook.presto.connector.thrift.api.PrestoThriftSplitBatch;
+import com.facebook.presto.connector.thrift.api.PrestoThriftTupleDomain;
+import com.facebook.presto.connector.thrift.location.HostLocationHandle;
+import com.facebook.presto.spi.HostAddress;
+import com.google.common.collect.ImmutableList;
+import com.google.common.util.concurrent.ListenableFuture;
+
+import java.util.List;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import static com.facebook.presto.spi.HostAddress.fromParts;
+
+public class MockConnectedThriftServiceProvider
+        implements ConnectedThriftServiceProvider
+{
+    private static final PrestoThriftServiceException RETRIABLE_EXCEPTION = new PrestoThriftServiceException("Mocking a retryable thrift failure", true);
+    private static final HostAddress DUMMY_HOST_ADDRESS = fromParts("test_host", 1234);
+    private static final HostLocationHandle DUMMY_HOST_LOCATION_HANDLE = new HostLocationHandle()
+    {
+        @Override
+        public HostAddress getHostAddress()
+        {
+            return DUMMY_HOST_ADDRESS;
+        }
+
+        @Override
+        public void reportFailure(Exception e) {}
+    };
+
+    private final AtomicInteger pendingFailuresToGenerate = new AtomicInteger();
+    private final AtomicInteger openServiceCount = new AtomicInteger();
+
+    public MockConnectedThriftServiceProvider() {}
+
+    public void setPendingFailuresToGenerate(int scheduledFailures)
+    {
+        this.pendingFailuresToGenerate.set(scheduledFailures);
+    }
+
+    public int getOpenServiceCount()
+    {
+        return openServiceCount.get();
+    }
+
+    @Override
+    public ConnectedThriftService anyHostClient()
+    {
+        openServiceCount.getAndIncrement();
+        return new MockConnectedThriftService(pendingFailuresToGenerate.getAndDecrement() > 0);
+    }
+
+    @Override
+    public ConnectedThriftService selectedHostClient(List<HostAddress> hosts)
+    {
+        openServiceCount.getAndIncrement();
+        return new MockConnectedThriftService(pendingFailuresToGenerate.getAndDecrement() > 0);
+    }
+
+    private final class MockConnectedThriftService
```

Comment:
static?

Reply:
It needs to update the `openServiceCount` when closing the thrift service.

Reply:
I removed the `failedService` member from this class and directly relied on the `pendingFailuresToGenerate` in the outer class.